extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,287 @@
+import {
+  RendererDef,
+  RendererInputDef,
+  ValidatorDef, // eslint-disable-line no-unused-vars
+  ValidatorOutputDef,
+  ValidatorResultType, // eslint-disable-line no-unused-vars
+} from '../../amp-a4a/0.1/a4a-render';
+import {dev} from '../../../src/log';
+import {utf8Encode} from '../../../src/utils/bytes';
+
+const TAG = 'amp-ad-network-base';
+
+export class AmpAdNetworkBase extends AMP.BaseElement {
+
+  constructor(element) {
+    super(element);
+
+    /** @private {Object<ValidatorResultType, !RendererDef>} */
+    this.boundRenderers_ = {};
+
+    /** @private {?ValidatorDef} */
+    this.boundValidator_ = null;
+
+    /** @private {?string} */
+    this.adUrl_ = null;
+
+    /** @private {?string} */
+    this.expandedAdUrl_ = null;
+
+    /** @private {?ArrayBuffer} */
+    this.unvalidatedBytes_ = null;
+
+    /** @private {!LayoutInfoDef} */
+    this.initialSize_ = {
+      // TODO(levitzky) handle non-numeric values.
+      width: element.getAttribute('width'),
+      height: element.getAttribute('height'),
+    };
+
+    /** @private {string} @const */
+    this.networkType_ = element.getAttribute('type') || 'anon';
+
+    /**
+     * @private {number} unique ID of the currently executing promise to allow
+     * for cancellation.
+     */
+    this.freshnessId_ = 0;
+
+    /** @private {function():boolean} a function that will return true if the
+     * freshness id changes due to unlayout. Initialized to always return
+     * false, and has an actual value set in onLayoutMeasure.
+     */
+    this.isStale_ = () => false;
+
+    /** {?AMP.AmpAdUIHandler} */
+    this.uiHandler = null;
+  }
+
+  /**
+   * @param {string} adUrl
+   * @protected
+   */
+  bindAdRequestUrl(adUrl) {
+    if (this.adUrl_) {
+      dev().warn(TAG, `Ad Request URL already bound: '${this.adUrl_}'`);
+    }
+    this.adUrl_ = adUrl;
+  }
+
+  /**
+   * @param {ValidatorResultType} resultType
+   * @param {!RendererDef} renderer
+   * @protected
+   */
+  bindRenderer(resultType, renderer) {
+    if (this.boundRenderers_[resultType]) {
+      dev().warn(TAG, `Rendering mode already bound for type '${resultType}'`);
+    }
+    this.boundRenderers_[resultType] = renderer;
+  }
+
+  /**
+   * @param {!ValidatorDef} validator
+   * @protected
+   */
+  bindValidator(validator) {
+    if (this.boundValidator_) {
+      dev().warn(TAG, 'Validator already bound.');
+    }
+    this.boundValidator_ = validator;
+  }
+
+  /**
+   * Called at various lifecycle stages. Can be overwitten by implementing
+   * networks to handle lifecycle event.
+   */
+  emitLifecycleEvent() {}
+
+  /**
+   * Collapses slot by setting its size to 0x0.
+   */
+  forceCollapse() {
+    dev().assert(this.uiHandler);
+    this.uiHandler.applyNoContentUI();
+    super.attemptChangeSize(0, 0);
+  }
+
+  /**
+   * Returns context data necessary to render creative in cross-domain
+   * NameFrame.
+   * @return {!JsonObject}
+   */
+  getCrossDomainContextData() {
```

Comment:
this is nameframe specific logic, should not belong to base class.  (it's not needed by all ad extensions)

Reply:
Done.